python/ray/data/block.py

```diff
@@ -477,29 +477,32 @@ def block_type(self) -> BlockType:
         raise NotImplementedError
 
 
-def _get_block_boundaries(block: Dict[str, np.ndarray]):
-    """Compute boundaries of the groups within a block.
-
-    The function expects the input block to be projected and sorted, i.e.,
-    only the grouping columns are present.
+def _get_block_boundaries(arrays: list[np.ndarray]) -> np.ndarray:
+    """Compute boundaries of the groups within a block, which is represented
+    by a list of sorted 1D numpy arrays for each column.
 
     Args:
-        block: a dict of numpy arrays with keys being the group column names.
-        This is generally coming from ``BlockAccessor.to_numpy()``.
+        arrays: a list of sorted 1D numpy arrays. This is generally given by the
+        dictionary values of ``BlockAccessor.to_numpy()``.
 
     Returns:
         A list of starting indices of each group and an end index of the last
         group, i.e., there are ``num_groups + 1`` entries and the first and last
         entries are 0 and ``len(array)`` respectively.
 
+    Note:
+        This function is implemented as finding the row indices where any of the
+        columns is not equal to the previous row.
     """
-    if len(block) == 1:
-        # For single key, just use the numpy array directly.
-        block = block.popitem()[1]
-    else:
-        # For multiple keys, we create a numpy record array to handle
-        # potentially different dtypes for each column.
-        dtype = [(k, v.dtype) for k, v in block.items()]
-        block = np.rec.fromarrays(list(block.values()), dtype=dtype)
-
-    return np.hstack([[0], np.where(block[1:] != block[:-1])[0] + 1, [len(block)]])
+    return np.hstack(
```

Comment:
Please add a comment elaborating on what you're trying to achieve (expression is not trivial)